NUSense/Core/Src/dynamixel/Chain.hpp

```diff
@@ -0,0 +1,208 @@
+
+            // Start the utility timer for the maximum timeout of 3 ms * 253 devices = 759 ms
+            utility_timer.begin(759);
+        };
+
+        /// @brief  Listen for a response from a broadcast ping to discover all devices on the chain
+        /// @note   This is a blocking function, and will wait for the full timeout of 759 ms
+        /// @retval All devices found on the chain
+        const std::vector<platform::NUSense::NUgus::ID>& discover_broadcast() {
+            // Only do discovery if we're currently discovering
+            if (!discovering) {
+                return devices;
+            }
+
+            // keep listening for packets until we timeout and there are no more packets
+            while (packet_handler.check_sts<3>(platform::NUSense::NUgus::ID::BROADCAST) != PacketHandler::Result::NONE
+                   || !utility_timer.has_timed_out()) {
+                switch (packet_handler.get_result()) {
+                    // If we got a good status, extract the ID
+                    // note: we need the braces for scoping the sts variable
+                    case PacketHandler::Result::SUCCESS: {
+                        auto sts = reinterpret_cast<const StatusReturnCommand<3>*>(packet_handler.get_sts_packet());
+                        // Add the ID to the chain
+                        devices.push_back(static_cast<platform::NUSense::NUgus::ID>(sts->id));
+                        if (sts->id <= static_cast<uint8_t>(platform::NUSense::NUgus::ID::MAX_SERVO_ID)) {
+                            servos.push_back(static_cast<platform::NUSense::NUgus::ID>(sts->id));
+                        }
+                        /// TODO: Potentially use the returned data to store the device model number and firmware
+                        /// version.
+                    } break;
+                    // If we got an error, hold onto it for logging
+                    case PacketHandler::Result::ERROR: {
+                        auto sts = reinterpret_cast<const StatusReturnCommand<3>*>(packet_handler.get_sts_packet());
+                        error_devices.push_back(static_cast<platform::NUSense::NUgus::ID>(sts->id));
+                    } break;
+                    // Skip the packet handler reset for a partial or none packet
+                    case PacketHandler::Result::PARTIAL:
+                    case PacketHandler::Result::NONE: continue;
+                    // For completeness
+                    default: break;
+                }
+                // Reset the packet handler for the next packet (that is likely already in the buffer)
+                // if we got a full packet (including an error), or a timeout.
+                packet_handler.ready();
+            };
+
+            // Unset flag
+            discovering = false;
+
+            return devices;
+        };
+
+        /// @brief  Gets all devices in the chain.
+        /// @retval A reference to the vector of devices in the chain.
```

Comment:
```suggestion
        /// @return A reference to the vector of devices in the chain.
```
Small nitpick: I am no expert on Doxygen, but I thought that `@retval` was for discrete named values like an enumeration or a boolean, and `@return` was for arbitrary values or references. See [this](https://sourceforge.net/p/doxygen/mailman/message/11651182/) and [this](https://micro-os-plus.github.io/develop/doxygen-style-guide/).